Fix accounts page (#1347)
polkadot-developers/polkadot-docs: smart-contracts/for-eth-devs/accounts.md

---
title: Accounts in Polkadot Hub Smart Contracts
description: Bridges Ethereum's 20-byte addresses with Polkadot's 32-byte accounts, enabling seamless interaction while maintaining compatibility with Ethereum tooling.
categories: Basics, Polkadot Protocol
---

# Accounts on the Polkadot Hub Smart Contracts

## Introduction

The Polkadot Hub natively utilizes Polkadot's 32-byte account system while providing interoperability with Ethereum's 20-byte addresses through an automatic conversion system. When interacting with smart contracts:

- Ethereum-compatible wallets (like MetaMask) can use their familiar 20-byte addresses.
- Polkadot accounts continue using their native 32-byte format.

> - The Polkadot Hub chain automatically handles conversion between the two formats behind the scenes:

Reviewer: This makes more sense to have "The" in front of "Polkadot Hub" cause you're referring to the chain explicitly

#### Limitations of Unmapped Accounts

Unmapped Substrate accounts (those created with Ed25519 or Sr25519 keypairs) face the following restrictions:

- **Cannot initiate transactions** through Ethereum-compatible interfaces (like MetaMask or other EVM-compatible wallets)
- **Cannot directly call smart contracts** using Ethereum RPC methods
- **Cannot transfer funds** to or from 20-byte Ethereum-compatible addresses without proper mapping
- **Limited interoperability** with Ethereum tooling and existing EVM infrastructure

The system can *receive* funds at the hashed 20-byte address derived from the 32-byte account, but the original account owner cannot control or access those funds through Ethereum-compatible methods without first establishing a proper mapping.

#### When Mapping is Required

Account mapping is **required** when:

1. You want to interact with smart contracts using Ethereum-compatible tools (MetaMask, Web3.js, Ethers.js)
2. You need to transfer funds using 20-byte address format
3. You want your existing Polkadot account to be accessible through EVM-compatible interfaces

> 4. You need bidirectional compatibility between Polkadot and Ethereum address formats

Reviewer: ```suggestion
- You want to interact with smart contracts using Ethereum-compatible tools (MetaMask, Web3.js, Ethers.js).
- You need to transfer funds using 20-byte address format.
- You want your existing Polkadot account to be accessible through EVM-compatible interfaces.
- You need bidirectional compatibility between Polkadot and Ethereum address formats.
```

These items didn't seem like they needed to be in sequential order

Account mapping is **not required** when:

- Using native Polkadot addresses (32-byte) exclusively with Substrate-native interfaces
- Interacting with parachains that don't use the Ethereum-compatible layer
- Using accounts that were originally created with secp256k1 keys (Ethereum-compatible from the start)

> 

Reviewer: ```suggestion
- Using native Polkadot addresses (32-byte) exclusively with Substrate-native interfaces.
- Interacting with parachains that don't use the Ethereum-compatible layer.
- Using accounts that were originally created with secp256k1 keys (Ethereum-compatible from the start).

```

### Account Mapping for Native Polkadot Accounts

If you have a native Polkadot account (32-byte format) that was created with a Polkadot/Substrate keypair (Ed25519/Sr25519) rather than an Ethereum-compatible keypair (secp256k1), you'll need to map your account to enable Ethereum compatibility.

To map your account, call the [`map_account`](https://paritytech.github.io/polkadot-sdk/master/pallet_revive/pallet/dispatchables/fn.map_account.html){target=\_blank} extrinsic of the [`pallet_revive`](https://paritytech.github.io/polkadot-sdk/master/pallet_revive/index.html){target=\_blank} pallet using your original Substrate account. This creates a stateful mapping that allows your 32-byte account to interact with the Ethereum-compatible smart contract system.

**Mapping Process:**

1. **Call the extrinsic**: Use your Substrate wallet to call `pallet_revive.map_account()`
2. **Pay the deposit**: A deposit is required and held while the mapping exists (refundable upon unmapping)

> 3. **Receive confirmation**: Once mapped, your account can be used with both Polkadot and Ethereum interfaces

Reviewer: ```suggestion
1. **Call the extrinsic**: Use your Substrate wallet to call `pallet_revive.map_account()`.
2. **Pay the deposit**: A deposit is required and held while the mapping exists (refundable upon unmapping).
3. **Receive confirmation**: Once mapped, your account can be used with both Polkadot and Ethereum interfaces.
```

The fallback mechanism is integrated into the [`to_account_id`](https://paritytech.github.io/polkadot-sdk/master/pallet_revive/trait.AddressMapper.html#tymethod.to_account_id){target=\_blank} function. It provides a safety net for address conversion by:

- First, attempting to retrieve stored mapping data from [`OriginalAccount`](https://paritytech.github.io/polkadot-sdk/master/pallet_revive/pallet/storage_types/struct.OriginalAccount.html){target=\_blank} storage
- Falling back to the default conversion method (Keccak-256 hash) if no explicit mapping exists

> - Maintaining consistency in address representation across the system

Reviewer: ```suggestion
- First, attempting to retrieve stored mapping data from [`OriginalAccount`](https://paritytech.github.io/polkadot-sdk/master/pallet_revive/pallet/storage_types/struct.OriginalAccount.html){target=\_blank} storage.
- Falling back to the default conversion method (Keccak-256 hash) if no explicit mapping exists.
- Maintaining consistency in address representation across the system.
```

### How Fallback Works with Unmapped Accounts

When an unmapped 32-byte Polkadot account needs to be represented as a 20-byte Ethereum address, the system:

1. **Checks for explicit mapping**: First looks in the `OriginalAccount` storage to see if the account has been explicitly mapped via `map_account`
2. **Applies fallback conversion**: If no mapping exists, automatically converts the 32-byte account to a 20-byte address by:
   - Hashing the full 32-byte account with Keccak-256
   - Taking the last 20 bytes of the resulting hash

> 3. **Uses the derived address**: This fallback address can receive funds, but the account owner cannot spend those funds through Ethereum-compatible interfaces without explicit mapping

Reviewer: ```suggestion
1. **Checks for explicit mapping**: First looks in the `OriginalAccount` storage to see if the account has been explicitly mapped via `map_account`.
2. **Applies fallback conversion**: If no mapping exists, automatically converts the 32-byte account to a 20-byte address by:
   - Hashing the full 32-byte account with Keccak-256.
   - Taking the last 20 bytes of the resulting hash.
3. **Uses the derived address**: This fallback address can receive funds, but the account owner cannot spend those funds through Ethereum-compatible interfaces without explicit mapping.
```

**Important Considerations:**

- The fallback mechanism is **one-way for unmapped accounts** - while you can derive the 20-byte address from a 32-byte account, you cannot recover the original 32-byte account from the 20-byte hash without the stored mapping
- Funds sent to a fallback address of an unmapped account are not lost, but require explicit mapping to be accessible through Ethereum tools

> - For security and usability, it's recommended to establish explicit mappings rather than relying on fallback addresses for accounts that need Ethereum compatibility

Reviewer: ```suggestion
- The fallback mechanism is **one-way for unmapped accounts**. While you can derive the 20-byte address from a 32-byte account, you cannot recover the original 32-byte account from the 20-byte hash without the stored mapping.
- Funds sent to a fallback address of an unmapped account are not lost, but require explicit mapping to be accessible through Ethereum tools.
- For security and usability, it's recommended to establish explicit mappings rather than relying on fallback addresses for accounts that need Ethereum compatibility.
```
